synapse/handlers/message.py

```diff
@@ -175,10 +177,17 @@ def allowed(event_and_state):
 
             return True
 
-        events_and_states = filter(allowed, events_and_states)
+        event_and_state = filter(
+            allowed,
+            [
+                (e, event_id_to_state[e.event_id])
+                for e in events
+            ]
+        )
+
```

Comment:
This could be done using a single list comprehension:

    [event for event in events if allowed(event, event_id_to_state[event.event_id])]